### ISSUE TITLE
About eval_set and eval_metric

### ISSUE THREAD

**Issue by 995884191** [sb-ai-lab/Py-Boost#22]
Hello,

I hope this message finds you well. I’m very excited to learn from your model and appreciate your contribution to the community. I have a question regarding the pyboost library: does it have the capability to record losses for each round of tree training? Additionally, can I use eval_set and eval_metric to automatically track the intermediate loss values for both the training and validation sets?

Are there any related tutorials or documentation available for this feature? Thank you very much for your assistance!

**Comment by 995884191** [comment]
The code I am currently using can only produce two recording points: full_model.fit(train_X, train_y, eval_sets=[{'X': val_X, 'y': val_y}]). I kindly request your assistance.
![Uploading 微信图片_20241223170655.png…]()

**Comment by 995884191** [comment]
The code I am currently using can only produce two recording points: full_model.fit(train_X, train_y, eval_sets=[{'X': val_X, 'y': val_y}]). I kindly request your assistance.
![1734944812704](https://github.com/user-attachments/assets/d58f2504-3936-4bef-9b60-b3f2b77e9156)

**Comment by btbpanda** [comment]
Hi! Thanks for your feedback. There is no specific tutorial for this feature, but here is the information that should be helpful to you: 

1) Py-Boost tracks the evaluation metric only for the validation dataset. However, you can track multiple datasets for your experiment by passing a list of dicts as eval_sets: `eval_sets = [{'X': X_train, 'y': y_train}, {'X': val_X, 'y': val_y}]`. In this case both datasets will be used to compute the metric, but only the last one will be used for early stopping criteria. 

2) Full evaluation history is available in `model.history`. It is a nested list where `model.history[i][j]` is the metric value at iteration i dataset j

3) Only a single metric is now available for tracking. The loss function used for optimization is not computed by itself, for GBDT only gradients and hessians are reqired, so, it is not tracked also. The only exception is the case when the loss matches the metric, for example, crossentropy loss and crossentropy metric. But anyway, you should keep in mind that for py-boost they are different entities - loss is crossentropy gradient, but metric is crossentropy by itself

4) If you really need to track multiple metrics, it is possible to do via creating your custom callback. You can check the tutorial [here](https://github.com/sb-ai-lab/Py-Boost/blob/master/tutorials/Tutorial_3_Custom_features.ipynb). By doing that you will have an access to any intermediate compuatation result and thus calculate on fly anything you need for both train and valid datasets

**Comment by 995884191** [comment]
> Hi! Thanks for your feedback. There is no specific tutorial for this feature, but here is the information that should be helpful to you:
> 
> 1. Py-Boost tracks the evaluation metric only for the validation dataset. However, you can track multiple datasets for your experiment by passing a list of dicts as eval_sets: `eval_sets = [{'X': X_train, 'y': y_train}, {'X': val_X, 'y': val_y}]`. In this case both datasets will be used to compute the metric, but only the last one will be used for early stopping criteria.
> 2. Full evaluation history is available in `model.history`. It is a nested list where `model.history[i][j]` is the metric value at iteration i dataset j
> 3. Only a single metric is now available for tracking. The loss function used for optimization is not computed by itself, for GBDT only gradients and hessians are reqired, so, it is not tracked also. The only exception is the case when the loss matches the metric, for example, crossentropy loss and crossentropy metric. But anyway, you should keep in mind that for py-boost they are different entities - loss is crossentropy gradient, but metric is crossentropy by itself
> 4. If you really need to track multiple metrics, it is possible to do via creating your custom callback. You can check the tutorial [here](https://github.com/sb-ai-lab/Py-Boost/blob/master/tutorials/Tutorial_3_Custom_features.ipynb). By doing that you will have an access to any intermediate compuatation result and thus calculate on fly anything you need for both train and valid datasets

Thank you for taking the time to respond to my question! Your insights were incredibly helpful, and I truly appreciate it.
